File: tests/test_baselines.py
"""Bounded-orders / legality assertion on the scripted baselines.

Both baselines are pure functions of the world state, so this is a real
property check rather than a smoke test: for hundreds of pseudo-random
worlds, the directive they emit must validate against the same reply schema
an LLM's must, and the controls it compiles to must be in range.

Plus: the two baselines are *ordered* — `formation` beats `swarm` — which is
what gives the ladder a spread instead of a wall of draws.
"""

from __future__ import annotations

import math
import random

import pytest
from conftest import require_sim_wasm

from cogball import baselines, defaults
from cogball.control import Body, World, compile_controls


def random_world(rng: random.Random) -> World:
    ball = Body(rng.uniform(-20, 20), rng.uniform(-12.5, 12.5),
                rng.uniform(-30, 30), rng.uniform(-30, 30))
    robots = []
    for i in range(defaults.NUM_ROBOTS):
        angle = rng.uniform(-1.0, 1.0)
        norm = math.sqrt(angle * angle + 1.0)
        robots.append(Body(
            rng.uniform(-19.4, 19.4), rng.uniform(-11.9, 11.9),
            rng.uniform(-7, 7), rng.uniform(-7, 7),
            1.0 / norm, angle / norm, rng.uniform(-6, 6),
            rng.randint(0, 12)))
    return World(tick=rng.randint(0, 7200), ball=ball, robots=tuple(robots),
                 goals=(rng.randint(0, 5), rng.randint(0, 5)),
                 last_touch_robot=rng.randint(-1, 5),
                 last_touch_seat=rng.randint(-1, 1),
                 last_touch_tick=0, freeze_until=0)


def assert_legal(directive, seat: int) -> None:
    ids = defaults.robot_ids_for_seat(seat)
    assert len(directive.orders) == defaults.ROBOTS_PER_SEAT
    assert [o.robot_id for o in directive.orders] == list(ids)
    assert len(directive.note) <= defaults.NOTE_MAX_RUNES
    for order in directive.orders:
        assert order.role in defaults.ROLES
        assert order.intent in defaults.INTENTS
        assert order.kick in defaults.KICK_MODES
        assert order.pass_to is None or (
            order.pass_to in ids and order.pass_to != order.robot_id)
        x, y = order.target
        assert math.isfinite(x) and math.isfinite(y)
        assert -defaults.PITCH_X <= x <= defaults.PITCH_X
        assert -defaults.PITCH_Y <= y <= defaults.PITCH_Y
        assert len(order.say) <= defaults.SAY_MAX_RUNES


@pytest.mark.parametrize("name", defaults.BASELINES)
def test_baselines_emit_legal_directives_and_bounded_controls(name):
    rng = random.Random(2026_08_22)
    for _ in range(500):
        world = random_world(rng)
        directives = []
        for seat in range(defaults.SEATS):
            directive = baselines.scripted_directive(name, world, seat)
            assert_legal(directive, seat)
            directives.append(directive)
        ctl = compile_controls(world, directives)
        assert len(ctl) == defaults.NUM_ROBOTS * 3
        for i in range(defaults.NUM_ROBOTS):
            thrust = int.from_bytes(ctl[i * 3:i * 3 + 1], "big", signed=True)
            turn = int.from_bytes(ctl[i * 3 + 1:i * 3 + 2], "big", signed=True)
            assert -100 <= thrust <= 100
            assert -100 <= turn <= 100
            assert ctl[i * 3 + 2] in (0, 1)


def test_baselines_are_pure_functions_of_the_world():
    rng = random.Random(7)
    world = random_world(rng)
    for name in defaults.BASELINES:
        first = baselines.scripted_directive(name, world, 0)
        for _ in range(3):
            assert baselines.scripted_directive(name, world, 0) == first


def test_formation_always_posts_a_keeper():
    rng = random.Random(99)
    for _ in range(200):
        world = random_world(rng)
        for seat in range(defaults.SEATS):
            directive = baselines.formation(world, seat)
            roles = [o.role for o in directive.orders]
            assert roles.count("keeper") == 1


def test_swarm_sends_everyone_at_the_ball_in_the_other_half():
    world = World(
        tick=0, ball=Body(12.0, 0.0, 0.0, 0.0),
        robots=tuple(Body(-3.0 + i, 1.0 * i, 0.0, 0.0) for i in range(6)),
        goals=(0, 0), last_touch_robot=-1, last_touch_seat=-1,
        last_touch_tick=-1, freeze_until=0)
    directive = baselines.swarm(world, 0)   # Azure attacks +x: their half
    assert [o.intent for o in directive.orders] == ["chase"] * 3


@pytest.mark.slow
def test_formation_beats_swarm_over_a_full_match():
    """The baselines are ordered, so the ladder has a spread."""
    require_sim_wasm()
    from cogball.sim import CogballSim
    sim = CogballSim(seed=42, first_kickoff_seat=0)
    for _ in range(7200):
        world = World.from_state(sim.state())
        sim.set_controls(compile_controls(
            world, [baselines.formation(world, 0),
                    baselines.swarm(world, 1)]))
        sim.step()
    assert not sim.fault()
    assert sim.goals(0) > sim.goals(1), \
        f"formation {sim.goals(0)} - {sim.goals(1)} swarm"

File: tests/test_baselines.nim
## Bounded-orders / legality assertion on the scripted baselines, plus the
## round-1 corner regression, pinned.

import std/[random, unicode]
import lib/helpers

proc validate(sim: SimServer, seat: Seat, directive: Directive, what: string) =
  ## The reply schema, asserted against a produced directive. A baseline that
  ## cannot satisfy it is not a comparable policy.
  doAssert directive.note.runeLen <= MaxNoteRunes,
    what & ": note is " & $directive.note.runeLen & " runes"
  doAssert isValidUtf8(directive.note), what & ": note is not valid UTF-8"
  var seen: array[RobotsPerSeat, bool]
  for slot in 0 ..< RobotsPerSeat:
    let
      order = directive.robots[slot]
      index = firstRobotOf(seat) + slot
    doAssert not seen[slot]
    seen[slot] = true
    doAssert ord(order.role) in 0 .. ord(Role.high), what & ": bad role"
    doAssert ord(order.intent) in 0 .. ord(Intent.high), what & ": bad intent"
    doAssert ord(order.kick) in 0 .. ord(KickMode.high), what & ": bad kick"
    doAssert order.targetX >= CentreX - 20_000_000'i32 and
      order.targetX <= CentreX + 20_000_000'i32,
      what & ": target x " & $order.targetX & " is off the pitch"
    doAssert order.targetY >= CentreY - 12_500_000'i32 and
      order.targetY <= CentreY + 12_500_000'i32,
      what & ": target y " & $order.targetY & " is off the pitch"
    if order.passTo >= 0:
      doAssert seatOfRobot(int(order.passTo)) == seat,
        what & ": pass_to is an opponent"
      doAssert order.passTo != int32(index), what & ": pass_to is self"
    doAssert order.say.runeLen <= MaxSayRunes,
      what & ": say is " & $order.say.runeLen & " runes"
    doAssert isValidUtf8(order.say), what & ": say is not valid UTF-8"

proc boundedOrders() =
  ## 500 pseudo-random world states x both baselines: every emitted directive
  ## validates, and every compiled mask has only legal bits.
  const LegalBits = ButtonUp or ButtonDown or ButtonLeft or ButtonRight or
    ButtonSelect or ButtonA
  var rng = initRand(0xBA5E)
  var sim = playing(testConfig())
  for round in 0 ..< 500:
    sim.pseudoWorld(rng)
    for name in ["formation", "swarm"]:
      for seat in Seat:
        let directive = sim.baselineDirective(seat, name, round)
        sim.validate(seat, directive,
          name & " seat " & seatAlias(seat) & " round " & $round)
        sim.activeDirective[seat] = directive
        sim.hasDirective[seat] = true
      let masks = sim.compileMasks(sim.activeDirective)
      for i in 0 ..< RobotCount:
        doAssert (masks[i] and not LegalBits) == 0,
          name & ": illegal bit on " & robotId(i)
  report "500 states x both baselines produce legal, bounded orders"

proc unknownBaselineIsFormation() =
  var sim = playing(testConfig())
  let fallback = sim.baselineDirective(Azure, "not-a-baseline", 0)
  let formation = sim.formationDirective(Azure, 0)
  doAssert fallback.note == formation.note
  for slot in 0 ..< RobotsPerSeat:
    doAssert fallback.robots[slot].intent == formation.robots[slot].intent
  report "an unknown baseline name plays `formation`"

proc exactlyOneKeeper() =
  ## `formation` always fields exactly one keeper, and it is the robot nearest
  ## its own goal.
  var rng = initRand(11)
  var sim = playing(testConfig())
  for _ in 0 ..< 200:
    sim.pseudoWorld(rng)
    for seat in Seat:
      let directive = sim.formationDirective(seat, 0)
      var keepers = 0
      var keeperSlot = -1
      for slot in 0 ..< RobotsPerSeat:
        if directive.robots[slot].role == roleKeeper:
          inc keepers
          keeperSlot = slot
      doAssert keepers == 1,
        "formation fielded " & $keepers & " keepers for " & seatAlias(seat)
      let own = ownGoalX(seat)
      for slot in 0 ..< RobotsPerSeat:
        if slot == keeperSlot:
          continue
        doAssert abs(int64(sim.robots[firstRobotOf(seat) + keeperSlot].x) -
            int64(own)) <=
          abs(int64(sim.robots[firstRobotOf(seat) + slot].x) - int64(own)),
          "the keeper is not the deepest robot"
  report "`formation` always fields exactly one keeper, the deepest robot"

proc swarmRolesAreFixed() =
  ## `swarm` reports striker/striker/back: the deepest robot is the `back`
  ## whatever the ball is doing. The role used to flip to striker whenever the
  ## ball crossed into the opponent half, so the seat view's `last_role` and
  ## the feed described a robot that changed identity twice a minute while
  ## doing the same job.
  var rng = initRand(0x5A11)
  var sim = playing(testConfig())
  var sawBothHalves: array[2, bool]
  for _ in 0 ..< 200:
    sim.pseudoWorld(rng)
    for seat in Seat:
      let
        directive = sim.swarmDirective(seat, 0)
        deepest = sim.deepestRobot(seat) - firstRobotOf(seat)
      sawBothHalves[ord(sim.ballInOwnHalf(seat))] = true
      var backs = 0
      for slot in 0 ..< RobotsPerSeat:
        let role = directive.robots[slot].role
        if slot == deepest:
          doAssert role == roleBack,
            "swarm's deepest robot reported " & roleText(role) & ", not back"
          inc backs
        else:
          doAssert role == roleStriker,
            "swarm reported " & roleText(role) & " for a chasing robot"
      doAssert backs == 1, "swarm fielded " & $backs & " backs"
  doAssert sawBothHalves[0] and sawBothHalves[1],
    "the sweep never saw the ball in both halves, so it proved nothing"
  report "swarm's roles are striker/striker/back in both halves"

proc formationBeatsSwarm() =
  ## The pinned match: at seed 679961 a formation-vs-swarm episode completes,
  ## formation wins, and it is NOT 0-0 — the round-1 corner regression, where
  ## 6 of 20 scripted matches ended goalless stuck against the boards.
  let result = runScriptedMatch(testConfig(seed = 679961), "formation", "swarm")
  doAssert result.reason == reasonComplete,
    "the match did not complete: " & reasonText(result.reason)
  doAssert result.rule in {erFullTime, erMercy},
    "unexpected end rule " & endRuleText(result.rule)
  doAssert result.goals[Azure] + result.goals[Crimson] > 0,
    "the match ended 0-0 — the corner regression is back"
  doAssert result.goals[Azure] > result.goals[Crimson],
    "formation lost to swarm at the pinned seed: " &
      $result.goals[Azure] & "-" & $result.goals[Crimson]
  report "formation beats swarm at seed 679961, and not 0-0 (" &
    $result.goals[Azure] & "-" & $result.goals[Crimson] & ")"

proc goalsHappenAcrossSeeds() =
  ## Goals are not a fluke of one seed: a walled 3v3 pitch that produces
  ## goalless matches is not a game.
  var goalless = 0
  var total = 0
  for seed in [1, 7919, 15838, 23757, 31676, 679961]:
    let result = runScriptedMatch(testConfig(seed = seed), "formation", "swarm")
    inc total
    if result.goals[Azure] + result.goals[Crimson] == 0:
      inc goalless
  doAssert goalless == 0,
    $goalless & " of " & $total & " scripted matches ended goalless"
  report "no scripted match across six seeds ends goalless"

when isMainModule:
  echo "test_baselines"
  boundedOrders()
  unknownBaselineIsFormation()
  exactlyOneKeeper()
  swarmRolesAreFixed()
  formationBeatsSwarm()
  goalsHappenAcrossSeeds()
  echo "test_baselines: all good"
